web/js/codeworld_shared.js

```diff
@@ -122,95 +122,161 @@ var hintBlacklist = [
     "Pi",
 ];
 
-var codeWorldBuiltinsDocs = {};
+// codeWorldSymbols is variable containing annotations and documentation
+// of builtin and user-defined variables.
+// Expected format:
+// codeWorldSymbols = {
+//   codeWorldLogo: {
+//     declaration: "codeWorldLogo :: Picture",
+//     symbolStart: 0,
+//     symbolEnd: 13,
+//     doc: "The CodeWorld logo."
+//   }
+// }
+
+var codeWorldSymbols = {},
+    codeWorldBuiltinSymbols = {};
+
+function getWordStart(word, line) {
+    return line.indexOf(word);
+}
 
-function onHover(cm, data, node){
-    if (data && data.token && data.token.string) {
-        var token_name = data.token.string;
-        if (hintBlacklist.indexOf(token_name) == -1 && codeWorldBuiltinsDocs[token_name]){
-            return codeWorldBuiltinsDocs[token_name];
-        }
+function getWordEnd(word, line) {
+    var wordStart = getWordStart(word, line);
+    if (wordStart != -1) {
+        return wordStart + word.length
     }
-    return;
+    return -1;
 }
 
-// Hints and hover tooltips
-function registerStandardHints(successFunc)
-{
-    function createHint(line, wordStart, wordEnd, cname) {
-        var word = line.slice(wordStart, wordEnd);
-        if (!cname) cname = 'hint-word';
+function parseSymbolsFromCurrentCode() {
+    var lines = window.codeworldEditor.getValue().split('\n'),
+        word = null,
+        parseResults = {};
 
-        function renderer(elem, data, cur) {
-            if (wordStart > 0) {
-                elem.appendChild(document.createTextNode(line.slice(0, wordStart)));
+    lines.forEach(function(line) {
+        // f(x, y) =
+        if (/^\w+\(.*/.test(line)) {
+            word = line.split("(")[0];
+            parseResults[word] = {
+                symbolStart: getWordStart(word, line),
```

Comment:
What does it mean to populate `symbolStart` and `symbolEnd` without `declaration`?  I thought those were offsets into `declaration`.

Reply:
You are right, no meaning. Will delete.